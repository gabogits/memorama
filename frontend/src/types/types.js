export const types = {
  LOGIN: "LOGIN",
  LOGOUT: "LOGOUT",
  SET_CARDS_GAME: "SET_CARDS_GAME",
  RESET_CATEGORY: "RESET_CATEGORY",
  RESET: "RESET",
};
