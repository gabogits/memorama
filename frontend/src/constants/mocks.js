export const arrayImages = [
  {
    parid: "2JcixB1Ky3I",
    url: "https://images.unsplash.com/photo-1570824104453-508955ab713e?crop=entropy&cs=tinysrgb&fit=max&fm=jpg&ixid=MnwyMDM2Nzh8MHwxfHNlYXJjaHwxfHxjYXRzfGVufDB8MHwxfHwxNjQzODI1MDE0&ixlib=rb-1.2.1&q=80&w=400",
  },
  {
    parid: "46TvM-BVrRI",
    url: "https://images.unsplash.com/photo-1589883661923-6476cb0ae9f2?crop=entropy&cs=tinysrgb&fit=max&fm=jpg&ixid=MnwyMDM2Nzh8MHwxfHNlYXJjaHwyfHxjYXRzfGVufDB8MHwxfHwxNjQzODI1MDE0&ixlib=rb-1.2.1&q=80&w=400",
  },
  {
    parid: "YCPkW_r_6uA",
    url: "https://images.unsplash.com/photo-1517331156700-3c241d2b4d83?crop=entropy&cs=tinysrgb&fit=max&fm=jpg&ixid=MnwyMDM2Nzh8MHwxfHNlYXJjaHwzfHxjYXRzfGVufDB8MHwxfHwxNjQzODI1MDE0&ixlib=rb-1.2.1&q=80&w=400",
  },
  {
    parid: "mBRfYA0dYYE",
    url: "https://images.unsplash.com/photo-1583795128727-6ec3642408f8?crop=entropy&cs=tinysrgb&fit=max&fm=jpg&ixid=MnwyMDM2Nzh8MHwxfHNlYXJjaHw0fHxjYXRzfGVufDB8MHwxfHwxNjQzODI1MDE0&ixlib=rb-1.2.1&q=80&w=400",
  },
  {
    parid: "uy5t-CJuIK4",
    url: "https://images.unsplash.com/photo-1541781774459-bb2af2f05b55?crop=entropy&cs=tinysrgb&fit=max&fm=jpg&ixid=MnwyMDM2Nzh8MHwxfHNlYXJjaHw1fHxjYXRzfGVufDB8MHwxfHwxNjQzODI1MDE0&ixlib=rb-1.2.1&q=80&w=400",
  },
  {
    parid: "FbhNdD1ow2g",
    url: "https://images.unsplash.com/photo-1557246565-8a3d3ab5d7f6?crop=entropy&cs=tinysrgb&fit=max&fm=jpg&ixid=MnwyMDM2Nzh8MHwxfHNlYXJjaHw2fHxjYXRzfGVufDB8MHwxfHwxNjQzODI1MDE0&ixlib=rb-1.2.1&q=80&w=400",
  },
  {
    parid: "MNju0A6EeE0",
    url: "https://images.unsplash.com/photo-1559235038-1b0fadf76f78?crop=entropy&cs=tinysrgb&fit=max&fm=jpg&ixid=MnwyMDM2Nzh8MHwxfHNlYXJjaHw3fHxjYXRzfGVufDB8MHwxfHwxNjQzODI1MDE0&ixlib=rb-1.2.1&q=80&w=400",
  },
  {
    parid: "gSSK4u8yPpM",
    url: "https://images.unsplash.com/photo-1568043210943-0e8aac4b9734?crop=entropy&cs=tinysrgb&fit=max&fm=jpg&ixid=MnwyMDM2Nzh8MHwxfHNlYXJjaHw4fHxjYXRzfGVufDB8MHwxfHwxNjQzODI1MDE0&ixlib=rb-1.2.1&q=80&w=400",
  },
  {
    parid: "qs4j-39TaBQ",
    url: "https://images.unsplash.com/photo-1595511890410-3b8dc237a537?crop=entropy&cs=tinysrgb&fit=max&fm=jpg&ixid=MnwyMDM2Nzh8MHwxfHNlYXJjaHw5fHxjYXRzfGVufDB8MHwxfHwxNjQzODI1MDE0&ixlib=rb-1.2.1&q=80&w=400",
  },
  {
    parid: "21NRDbMJF94",
    url: "https://images.unsplash.com/photo-1574063413132-354db9f190fd?crop=entropy&cs=tinysrgb&fit=max&fm=jpg&ixid=MnwyMDM2Nzh8MHwxfHNlYXJjaHwxMHx8Y2F0c3xlbnwwfDB8MXx8MTY0MzgyNTAxNA&ixlib=rb-1.2.1&q=80&w=400",
  },
];
export const mockGameCards = [
  {
    parid: "gSSK4u8yPpM",
    url: "https://images.unsplash.com/photo-1568043210943-0e8aac4b9734?crop=entropy&cs=tinysrgb&fit=max&fm=jpg&ixid=MnwyMDM2Nzh8MHwxfHNlYXJjaHw4fHxjYXRzfGVufDB8MHwxfHwxNjQzODI1MDE0&ixlib=rb-1.2.1&q=80&w=400",
    nanoId: "UanozUmNaQ",
    flipped: false,
    active: true,
  },
  {
    parid: "2JcixB1Ky3I",
    url: "https://images.unsplash.com/photo-1570824104453-508955ab713e?crop=entropy&cs=tinysrgb&fit=max&fm=jpg&ixid=MnwyMDM2Nzh8MHwxfHNlYXJjaHwxfHxjYXRzfGVufDB8MHwxfHwxNjQzODI1MDE0&ixlib=rb-1.2.1&q=80&w=400",
    nanoId: "HSmfLRJf4b",
    flipped: false,
    active: true,
  },
  {
    parid: "uy5t-CJuIK4",
    url: "https://images.unsplash.com/photo-1541781774459-bb2af2f05b55?crop=entropy&cs=tinysrgb&fit=max&fm=jpg&ixid=MnwyMDM2Nzh8MHwxfHNlYXJjaHw1fHxjYXRzfGVufDB8MHwxfHwxNjQzODI1MDE0&ixlib=rb-1.2.1&q=80&w=400",
    nanoId: "feRFoRKQ6Z",
    flipped: false,
    active: true,
  },
  {
    parid: "mBRfYA0dYYE",
    url: "https://images.unsplash.com/photo-1583795128727-6ec3642408f8?crop=entropy&cs=tinysrgb&fit=max&fm=jpg&ixid=MnwyMDM2Nzh8MHwxfHNlYXJjaHw0fHxjYXRzfGVufDB8MHwxfHwxNjQzODI1MDE0&ixlib=rb-1.2.1&q=80&w=400",
    nanoId: "R6-xYjqtkH",
    flipped: false,
    active: true,
  },
  {
    parid: "FbhNdD1ow2g",
    url: "https://images.unsplash.com/photo-1557246565-8a3d3ab5d7f6?crop=entropy&cs=tinysrgb&fit=max&fm=jpg&ixid=MnwyMDM2Nzh8MHwxfHNlYXJjaHw2fHxjYXRzfGVufDB8MHwxfHwxNjQzODI1MDE0&ixlib=rb-1.2.1&q=80&w=400",
    nanoId: "9F7o3fBP29",
    flipped: false,
    active: true,
  },
  {
    parid: "MNju0A6EeE0",
    url: "https://images.unsplash.com/photo-1559235038-1b0fadf76f78?crop=entropy&cs=tinysrgb&fit=max&fm=jpg&ixid=MnwyMDM2Nzh8MHwxfHNlYXJjaHw3fHxjYXRzfGVufDB8MHwxfHwxNjQzODI1MDE0&ixlib=rb-1.2.1&q=80&w=400",
    nanoId: "F6jx1v-Vyd",
    flipped: false,
    active: true,
  },
  {
    parid: "gSSK4u8yPpM",
    url: "https://images.unsplash.com/photo-1568043210943-0e8aac4b9734?crop=entropy&cs=tinysrgb&fit=max&fm=jpg&ixid=MnwyMDM2Nzh8MHwxfHNlYXJjaHw4fHxjYXRzfGVufDB8MHwxfHwxNjQzODI1MDE0&ixlib=rb-1.2.1&q=80&w=400",
    nanoId: "oLzJktU-Az",
    flipped: false,
    active: true,
  },
  {
    parid: "FbhNdD1ow2g",
    url: "https://images.unsplash.com/photo-1557246565-8a3d3ab5d7f6?crop=entropy&cs=tinysrgb&fit=max&fm=jpg&ixid=MnwyMDM2Nzh8MHwxfHNlYXJjaHw2fHxjYXRzfGVufDB8MHwxfHwxNjQzODI1MDE0&ixlib=rb-1.2.1&q=80&w=400",
    nanoId: "P9hCUD77_p",
    flipped: false,
    active: true,
  },
  {
    parid: "YCPkW_r_6uA",
    url: "https://images.unsplash.com/photo-1517331156700-3c241d2b4d83?crop=entropy&cs=tinysrgb&fit=max&fm=jpg&ixid=MnwyMDM2Nzh8MHwxfHNlYXJjaHwzfHxjYXRzfGVufDB8MHwxfHwxNjQzODI1MDE0&ixlib=rb-1.2.1&q=80&w=400",
    nanoId: "MUa5PByAj8",
    flipped: false,
    active: true,
  },
  {
    parid: "YCPkW_r_6uA",
    url: "https://images.unsplash.com/photo-1517331156700-3c241d2b4d83?crop=entropy&cs=tinysrgb&fit=max&fm=jpg&ixid=MnwyMDM2Nzh8MHwxfHNlYXJjaHwzfHxjYXRzfGVufDB8MHwxfHwxNjQzODI1MDE0&ixlib=rb-1.2.1&q=80&w=400",
    nanoId: "1zTBC5WcO-",
    flipped: false,
    active: true,
  },
  {
    parid: "mBRfYA0dYYE",
    url: "https://images.unsplash.com/photo-1583795128727-6ec3642408f8?crop=entropy&cs=tinysrgb&fit=max&fm=jpg&ixid=MnwyMDM2Nzh8MHwxfHNlYXJjaHw0fHxjYXRzfGVufDB8MHwxfHwxNjQzODI1MDE0&ixlib=rb-1.2.1&q=80&w=400",
    nanoId: "agIioCJ5_t",
    flipped: false,
    active: true,
  },
  {
    parid: "uy5t-CJuIK4",
    url: "https://images.unsplash.com/photo-1541781774459-bb2af2f05b55?crop=entropy&cs=tinysrgb&fit=max&fm=jpg&ixid=MnwyMDM2Nzh8MHwxfHNlYXJjaHw1fHxjYXRzfGVufDB8MHwxfHwxNjQzODI1MDE0&ixlib=rb-1.2.1&q=80&w=400",
    nanoId: "V02rFiPXi_",
    flipped: false,
    active: true,
  },
  {
    parid: "21NRDbMJF94",
    url: "https://images.unsplash.com/photo-1574063413132-354db9f190fd?crop=entropy&cs=tinysrgb&fit=max&fm=jpg&ixid=MnwyMDM2Nzh8MHwxfHNlYXJjaHwxMHx8Y2F0c3xlbnwwfDB8MXx8MTY0MzgyNTAxNA&ixlib=rb-1.2.1&q=80&w=400",
    nanoId: "4qzwACkR6F",
    flipped: false,
    active: true,
  },
  {
    parid: "qs4j-39TaBQ",
    url: "https://images.unsplash.com/photo-1595511890410-3b8dc237a537?crop=entropy&cs=tinysrgb&fit=max&fm=jpg&ixid=MnwyMDM2Nzh8MHwxfHNlYXJjaHw5fHxjYXRzfGVufDB8MHwxfHwxNjQzODI1MDE0&ixlib=rb-1.2.1&q=80&w=400",
    nanoId: "lWOOBcAgTf",
    flipped: false,
    active: true,
  },
  {
    parid: "MNju0A6EeE0",
    url: "https://images.unsplash.com/photo-1559235038-1b0fadf76f78?crop=entropy&cs=tinysrgb&fit=max&fm=jpg&ixid=MnwyMDM2Nzh8MHwxfHNlYXJjaHw3fHxjYXRzfGVufDB8MHwxfHwxNjQzODI1MDE0&ixlib=rb-1.2.1&q=80&w=400",
    nanoId: "vJG_OfRr0p",
    flipped: false,
    active: true,
  },
  {
    parid: "46TvM-BVrRI",
    url: "https://images.unsplash.com/photo-1589883661923-6476cb0ae9f2?crop=entropy&cs=tinysrgb&fit=max&fm=jpg&ixid=MnwyMDM2Nzh8MHwxfHNlYXJjaHwyfHxjYXRzfGVufDB8MHwxfHwxNjQzODI1MDE0&ixlib=rb-1.2.1&q=80&w=400",
    nanoId: "70ZYeFj5sc",
    flipped: false,
    active: true,
  },
  {
    parid: "46TvM-BVrRI",
    url: "https://images.unsplash.com/photo-1589883661923-6476cb0ae9f2?crop=entropy&cs=tinysrgb&fit=max&fm=jpg&ixid=MnwyMDM2Nzh8MHwxfHNlYXJjaHwyfHxjYXRzfGVufDB8MHwxfHwxNjQzODI1MDE0&ixlib=rb-1.2.1&q=80&w=400",
    nanoId: "HTbumxhoHL",
    flipped: false,
    active: true,
  },
  {
    parid: "21NRDbMJF94",
    url: "https://images.unsplash.com/photo-1574063413132-354db9f190fd?crop=entropy&cs=tinysrgb&fit=max&fm=jpg&ixid=MnwyMDM2Nzh8MHwxfHNlYXJjaHwxMHx8Y2F0c3xlbnwwfDB8MXx8MTY0MzgyNTAxNA&ixlib=rb-1.2.1&q=80&w=400",
    nanoId: "sD2Y-y1-jT",
    flipped: false,
    active: true,
  },
  {
    parid: "2JcixB1Ky3I",
    url: "https://images.unsplash.com/photo-1570824104453-508955ab713e?crop=entropy&cs=tinysrgb&fit=max&fm=jpg&ixid=MnwyMDM2Nzh8MHwxfHNlYXJjaHwxfHxjYXRzfGVufDB8MHwxfHwxNjQzODI1MDE0&ixlib=rb-1.2.1&q=80&w=400",
    nanoId: "fctB2ZTC44",
    flipped: false,
    active: true,
  },
  {
    parid: "qs4j-39TaBQ",
    url: "https://images.unsplash.com/photo-1595511890410-3b8dc237a537?crop=entropy&cs=tinysrgb&fit=max&fm=jpg&ixid=MnwyMDM2Nzh8MHwxfHNlYXJjaHw5fHxjYXRzfGVufDB8MHwxfHwxNjQzODI1MDE0&ixlib=rb-1.2.1&q=80&w=400",
    nanoId: "krA7fPHREn",
    flipped: false,
    active: true,
  },
];
export const players = {
  numberPlayers: 2,
  dataPlayers: [
    { avatar: "witch", id: 1, name: "juan" },
    { avatar: "dracula", id: 2, name: "bety" },
  ],
};
