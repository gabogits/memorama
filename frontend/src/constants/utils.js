export const playersButtons = [
  { number: 1, selected: false, label: "Un jugador" },
  { number: 2, selected: false, label: "Dos jugadores" },
];
export const avatars = [
  { id: 1, name: "witch", url: require("../assets/images/avatar/witch.png") },
  {
    id: 2,
    name: "dracula",
    url: require("../assets/images/avatar/dracula.png"),
  },
  {
    id: 3,
    name: "frankenstein",
    url: require("../assets/images/avatar/frankenstein.png"),
  },
  { id: 4, name: "scream", url: require("../assets/images/avatar/scream.png") },
  {
    id: 5,
    name: "werewolf",
    url: require("../assets/images/avatar/werewolf.png"),
  },
  { id: 6, name: "zombie", url: require("../assets/images/avatar/zombie.png") },
];

export const MAX_SCORE = 10000;
export const MAX_TIME_GAME = 300;

export const NUMBER_CARDS = 10;

export const url = "https://api.unsplash.com/search/photos?query=";
