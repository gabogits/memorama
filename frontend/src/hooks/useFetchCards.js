import { useState } from "react";

export const useFetchCards = (url, NUMBER_CARDS) => {
  const [arrayImages, setArrayImages] = useState([]);
  const [error, setError] = useState("");
  const [loading, setLoading] = useState(false);

  const handlefetchCards = async (query) => {
    setLoading(true);
    await fetch(
      `${url}${query}&orientation=landscape&client_id=2CIGTbb0j0WvgJ_TI2k0fGeJ-YjtmTAthvAwgX4ytZE&per_page=${NUMBER_CARDS}&order_by=popular`
    )
      .then((resp) => resp.json())
      .then(({ results }) => {
        const arrayImages = results.map((item) => {
          return {
            parid: item.id,
            url: item.urls.small,
          };
        });

        if (arrayImages.length === NUMBER_CARDS) {
          setArrayImages(arrayImages);
          setError("");
        } else {
          setArrayImages([]);
          setError("No se encontraton resultados, intenta con otro tema");
        }
        setLoading(false);
      })
      .catch(() => {
        setError("Ocurrio un error");
        setLoading(false);
      });
  };

  return {
    arrayImages,
    loading,
    error,
    setArrayImages,
    handlefetchCards,
  };
};
