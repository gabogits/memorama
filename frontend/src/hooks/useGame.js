import { useState, useEffect, useRef } from "react";
import { nanoid } from "nanoid";
import { MAX_SCORE, MAX_TIME_GAME } from "../constants/utils";
import { calculateScore } from "../helpers/calculateScore";
export const useGame = (
  numberPlayers,
  dataPlayers,
  arrayImages,
  handleStart,
  handlePause,
  handleReset,
  timer
) => {
  const isMounted = useRef(true);
  const [cardsGame, setCardsGame] = useState([]);
  const [endGame, setEndGame] = useState(false);
  const [score, setScore] = useState(null);

  const [move, setMove] = useState({
    playerMove: dataPlayers[0],
    cardsMove: [],
  });
  const { playerMove, cardsMove } = move;
  const [gameData, setGameData] = useState([]);

  const handleCardFlip = (id, parid, active) => {
    const cardAlreadyFlipped = cardsMove.find((elem) => elem.id === id);
    if (!cardAlreadyFlipped && active && cardsMove.length < 2) {
      setMove({
        ...move,
        cardsMove: [...cardsMove, { parid, id }],
      });

      const newCardsGame = cardsGame.map((item) =>
        item.nanoId === id ? { ...item, flipped: true } : item
      );

      setCardsGame(newCardsGame);
    }
  };

  const initGame = () => {
    handleReset();
    setEndGame(false);
    const cardsGame = [...arrayImages, ...arrayImages]
      .map((item) => {
        return {
          ...item,
          nanoId: nanoid(10),
          flipped: false,
          active: true,
        };
      })
      .sort(() => Math.random() - 0.5);

    setCardsGame(cardsGame);

    setMove({
      playerMove: dataPlayers[0],
      cardsMove: [],
    });

    setGameData(
      dataPlayers.map((item) => {
        return {
          ...item,
          score: 0,
        };
      })
    );
    setScore(null);
    handleStart();
  };

  const cleanGameBoard = () => {
    let newCardsGame;
    let newDataGame;

    setTimeout(() => {
      let playerMoveData;
      if (cardsMove[0].parid === cardsMove[1].parid) {
        newCardsGame = cardsGame.map((item) =>
          item.parid === cardsMove[0].parid ? { ...item, active: false } : item
        );
        newDataGame = gameData.map((item) =>
          item.id === playerMove.id ? { ...item, score: item.score + 1 } : item
        );

        setGameData(newDataGame);

        const activeCards = newCardsGame.filter((item) => item.active);
        if (activeCards.length === 0) {
          handlePause();
          setScore(calculateScore(MAX_TIME_GAME, timer, MAX_SCORE));
          setEndGame(true);
        }
      } else {
        newCardsGame = cardsGame.map((item) =>
          item.flipped ? { ...item, flipped: false } : item
        );
        if (numberPlayers === 2) {
          playerMoveData = {
            playerMove: dataPlayers.find((item) => item.id !== playerMove.id),
          };
        }
      }
      setMove({
        ...move,
        cardsMove: [],
        ...playerMoveData,
      });

      setCardsGame(newCardsGame);
    }, 1200);
  };
  useEffect(() => {
    if (isMounted.current) {
      initGame();
    }
    return () => {
      handlePause();
      handleReset();
      isMounted.current = false;
    };
    // eslint-disable-next-line
  }, []);
  useEffect(() => {
    if (isMounted.current && cardsMove.length === 2) {
      cleanGameBoard();
    }
    // eslint-disable-next-line
  }, [cardsMove]);

  return {
    playerMove,
    cardsGame,
    gameData,
    endGame,
    score,
    initGame,
    handleCardFlip,
  };
};
