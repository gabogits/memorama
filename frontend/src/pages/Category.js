import React, { useContext, useState } from "react";
import Loader from "../components/templates/Loader";
import { useFetchCards } from "../hooks/useFetchCards";
import { useForm } from "../hooks/useForm";
import GameContext from "../context/GameContext";
import { types } from "../types/types";
import { useHistory } from "react-router-dom";
import Layout from "../components/layout/Layout";
import CardsPreview from "../components/CardsPreview";
import Error from "../components/templates/Error";
import { NUMBER_CARDS, url } from "../constants/utils";

const initialState = {
  categoryCombo: "",
  categoryCustom: "",
};
const Category = () => {
  const { dispatch } = useContext(GameContext);
  const history = useHistory();
  const [values, handleInputChange] = useForm(initialState);
  const { arrayImages, loading, error, setArrayImages, handlefetchCards } =
    useFetchCards(url, NUMBER_CARDS);

  const { categoryCombo, categoryCustom } = values;

  const [showCards, setShowCards] = useState(false);

  const handleChange = (e) => {
    handleInputChange(e);
    validationCategory(e);
  };
  const validationCategory = ({ target }) => {
    setShowCards(false);
    setArrayImages([]);
    if (
      target.name === "categoryCombo" &&
      target.value &&
      target.value !== "other"
    ) {
      handlefetchCards(target.value);
    }
  };

  const handleSubmit = (e) => {
    e.preventDefault();

    dispatch({
      type: types.SET_CARDS_GAME,
      payload: {
        category: categoryCombo !== "other" ? categoryCombo : categoryCustom,
        arrayImages,
      },
    });
    history.push("/juego");
  };

  return (
    <Layout>
      <section className="form top-bottom-space">
        <div className="container">
          <section className="section-format col">
            <form>
              <div className="field ">
                <label>Selecciona un tema</label>
                <div className="select-simple">
                  <div className="select-simple__container">
                    <select
                      name="categoryCombo"
                      onChange={(e) => handleChange(e)}
                      value={categoryCombo}
                    >
                      <option value="">Seleccionar</option>
                      <option value="cats">Gatos</option>
                      <option value="dogs">Perros</option>
                      <option value="other">Otro</option>
                    </select>
                  </div>
                </div>
              </div>
              {categoryCombo === "other" && (
                <div className="field input-category-custom">
                  <label>Nombre de tema</label>
                  <div className="input-combo-icon">
                    <input
                      type="text"
                      className="u-full-width"
                      placeholder="Por ejemplo Disney, Skate, New York, Coches, etc"
                      name="categoryCustom"
                      value={categoryCustom}
                      onChange={(e) => handleChange(e)}
                    />
                    <button
                      type="submit"
                      className="btn-icon btn-absolute btn-color-5 icon-search"
                      value="enviar"
                      onClick={(e) => {
                        e.preventDefault();
                        handlefetchCards(categoryCustom);
                      }}
                    ></button>
                  </div>
                </div>
              )}

              {loading && <Loader />}
              {error && <Error msg={error} />}
              {arrayImages.length === NUMBER_CARDS && (
                <>
                  <div className="buttons-row">
                    <button
                      type="submit"
                      className="btn-primary show-cards-btn  btn-size-2 btn-color-2"
                      value="enviar"
                      onClick={(e) => {
                        e.preventDefault();
                        setShowCards(true);
                      }}
                    >
                      Ver las tarjetas
                    </button>

                    <button
                      type="submit"
                      className="btn-primary start-game-btn btn-size-2 btn-color-4"
                      value="enviar"
                      onClick={handleSubmit}
                    >
                      Empezar el juego
                    </button>
                  </div>
                </>
              )}
            </form>
          </section>

          {showCards && (
            <CardsPreview
              setShowCards={setShowCards}
              arrayImages={arrayImages}
            />
          )}
        </div>
      </section>
    </Layout>
  );
};

export default Category;
