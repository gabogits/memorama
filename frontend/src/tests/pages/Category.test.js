import { mount } from "enzyme";
import { Router } from "react-router-dom";
import GameContext from "../../context/GameContext";
import Category from "../../pages/Category";
import { useFetchCards } from "../../hooks/useFetchCards";
import { arrayImages } from "../../constants/mocks";
import { types } from "../../types/types";

jest.mock("../../hooks/useFetchCards");

describe("Pruebas en <Category />", () => {
  let historyMock = {
    push: jest.fn(),
    replace: jest.fn(),
    location: jest.fn(),
    listen: jest.fn(),
    createHref: jest.fn(),
  };
  const contextValue = {
    dispatch: jest.fn(),
    state: {
      numberPlayers: 0,
      dataPlayers: [],
      arrayImages: [],
      category: null,
    },
  };

  afterEach(() => {
    jest.clearAllMocks();
  });

  test("Debe de mostarrse correctamente el snapshot", () => {
    useFetchCards.mockReturnValue({
      arrayImages: [],
      loading: false,
      error: "",
      setArrayImages: jest.fn(),
      handlefetchCards: jest.fn(),
    });

    const wrapper = mount(
      <GameContext.Provider value={contextValue}>
        <Router history={historyMock}>
          <Category />
        </Router>
      </GameContext.Provider>
    );
    expect(wrapper.find("CardsPreview").exists()).toBe(false);
    expect(wrapper.find("Loader").exists()).toBe(false);
    expect(wrapper.find("Error").exists()).toBe(false);
    expect(wrapper.find(".input-category-custom").exists()).toBe(false);

    expect(wrapper.find(".show-cards-btn").exists()).toBe(false);
    expect(wrapper.find(".start-game-btn").exists()).toBe(false);
    expect(wrapper.find("CardsPreview").exists()).toBe(false);
    expect(wrapper).toMatchSnapshot();
  });
  test("No Debe de mostarrse el input categoryCustom cuando elijas cats o dogs, debe de mostrarse el loader y los botones mostrar tarjetas y empezar juego no deben aparecer aÃ¹n", () => {
    useFetchCards.mockReturnValue({
      arrayImages: [],
      loading: true,
      error: "",
      setArrayImages: jest.fn(),
      handlefetchCards: jest.fn(),
    });

    const wrapper = mount(
      <GameContext.Provider value={contextValue}>
        <Router history={historyMock}>
          <Category />
        </Router>
      </GameContext.Provider>
    );
    const e = { target: { value: "cats", name: "categoryCombo" } };
    wrapper.find("select").simulate("change", e);
    expect(wrapper.find(".input-category-custom").exists()).toBe(false);
    expect(wrapper.find("Loader").exists()).toBe(true);
    expect(wrapper.find(".show-cards-btn").exists()).toBe(false);
    expect(wrapper.find(".start-game-btn").exists()).toBe(false);
  });

  test("Debe de mostrar los botones para ver tarjetas y empezar el juego, se oculta el loader", () => {
    useFetchCards.mockReturnValue({
      arrayImages,
      loading: false,
      error: "",
      setArrayImages: jest.fn(),
      handlefetchCards: jest.fn(),
    });

    const wrapper = mount(
      <GameContext.Provider value={contextValue}>
        <Router history={historyMock}>
          <Category />
        </Router>
      </GameContext.Provider>
    );
    const e = { target: { value: "cats", name: "categoryCombo" } };
    wrapper.find("select").simulate("change", e);
    expect(wrapper.find("Loader").exists()).toBe(false);
    expect(wrapper.find(".show-cards-btn").exists()).toBe(true);
    expect(wrapper.find(".start-game-btn").exists()).toBe(true);
    expect(wrapper.find("CardsPreview").exists()).toBe(false);
  });
  test("Debe de mostrar las tarjetas al darle click al boton Ver las tarjetas", () => {
    useFetchCards.mockReturnValue({
      arrayImages,
      loading: false,
      error: "",
      setArrayImages: jest.fn(),
      handlefetchCards: jest.fn(),
    });

    const wrapper = mount(
      <GameContext.Provider value={contextValue}>
        <Router history={historyMock}>
          <Category />
        </Router>
      </GameContext.Provider>
    );
    wrapper.find(".show-cards-btn").simulate("click");
    expect(wrapper.find("CardsPreview").exists()).toBe(true);
    expect(wrapper.find("Card").length).toBe(10);
  });

  test("Debe de empezar el juego con categoria predefinida", () => {
    useFetchCards.mockReturnValue({
      arrayImages,
      loading: false,
      error: "",
      setArrayImages: jest.fn(),
      handlefetchCards: jest.fn(),
    });

    const contextNewValue = {
      type: types.SET_CARDS_GAME,
      payload: {
        category: "cats",
        arrayImages,
      },
    };

    const wrapper = mount(
      <GameContext.Provider value={contextValue}>
        <Router history={historyMock}>
          <Category />
        </Router>
      </GameContext.Provider>
    );
    const e = { target: { value: "cats", name: "categoryCombo" } };
    wrapper.find("select").simulate("change", e);

    wrapper.find(".start-game-btn").simulate("click");
    expect(contextValue.dispatch).toHaveBeenCalledTimes(1);
    expect(contextValue.dispatch).toHaveBeenCalledWith(contextNewValue);
    expect(historyMock.push).toHaveBeenCalledWith("/juego");
  });

  test("Debe de mostarrse el input categoryCustom cuando elijas other", () => {
    useFetchCards.mockReturnValue({
      arrayImages: [],
      loading: true,
      error: "",
      setArrayImages: jest.fn(),
      handlefetchCards: jest.fn(),
    });

    const wrapper = mount(
      <GameContext.Provider value={contextValue}>
        <Router history={historyMock}>
          <Category />
        </Router>
      </GameContext.Provider>
    );
    const e = { target: { value: "other", name: "categoryCombo" } };

    wrapper.find("select").simulate("change", e);
    const d = { target: { value: "Harry potter", name: "categoryCustom" } };
    wrapper.find(".input-category-custom input").simulate("change", d);
    expect(wrapper.find(".input-category-custom").exists()).toBe(true);
    expect(wrapper.find("Loader").exists()).toBe(true);
  });

  test("Debe de mostarrse el error", () => {
    useFetchCards.mockReturnValue({
      arrayImages: [],
      loading: false,
      error: "No se encontraton resultados, intenta con otro tema",
      setArrayImages: jest.fn(),
      handlefetchCards: jest.fn(),
    });

    const wrapper = mount(
      <GameContext.Provider value={contextValue}>
        <Router history={historyMock}>
          <Category />
        </Router>
      </GameContext.Provider>
    );
    const e = { target: { value: "other", name: "categoryCombo" } };

    wrapper.find("select").simulate("change", e);
    const d = { target: { value: "Harryss", name: "categoryCustom" } };
    wrapper.find(".input-category-custom input").simulate("change", d);
    expect(wrapper.find(".input-category-custom").exists()).toBe(true);
    expect(wrapper.find("Loader").exists()).toBe(false);
    expect(wrapper.find("Error").exists()).toBe(true);
    expect(wrapper.find(".alert").text().trim()).toBe(
      "No se encontraton resultados, intenta con otro tema"
    );
  });

  test("Empezar juego con categoria customizada", () => {
    useFetchCards.mockReturnValue({
      arrayImages,
      loading: false,
      error: "",
      setArrayImages: jest.fn(),
      handlefetchCards: jest.fn(),
    });

    const contextNewValue = {
      type: types.SET_CARDS_GAME,
      payload: {
        category: "Harry potter",
        arrayImages,
      },
    };

    const wrapper = mount(
      <GameContext.Provider value={contextValue}>
        <Router history={historyMock}>
          <Category />
        </Router>
      </GameContext.Provider>
    );
    const e = { target: { value: "other", name: "categoryCombo" } };

    wrapper.find("select").simulate("change", e);
    const d = { target: { value: "Harry potter", name: "categoryCustom" } };
    wrapper.find(".input-category-custom input").simulate("change", d);
    wrapper.find(".start-game-btn").simulate("click");
    expect(contextValue.dispatch).toHaveBeenCalledTimes(1);
    expect(contextValue.dispatch).toHaveBeenCalledWith(contextNewValue);
    expect(historyMock.push).toHaveBeenCalledWith("/juego");
  });
});
