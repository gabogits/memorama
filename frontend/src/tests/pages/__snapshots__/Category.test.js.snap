// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <Category /> Debe de mostarrse correctamente el snapshot 1`] = `
<main>
  <header>
    <div
      className=" container container-header"
    >
      <div
        className="logo"
      >
        <figure>
          <a
            onClick={[Function]}
          >
            <span
              className="logo-desk"
            >
              MEMORAMA
            </span>
            <span
              className="logo-mobile"
            >
              M
            </span>
             
          </a>
        </figure>
      </div>
      <nav>
        <div
          className="nav-left"
        />
        <div
          className="nav-right"
        >
          <div
            className="nav-list"
          >
            <ul>
              <li>
                <button
                  onClick={[Function]}
                >
                  Inicio
                </button>
              </li>
              <li>
                <a
                  aria-current={null}
                  onClick={[Function]}
                >
                  Ranking
                </a>
              </li>
            </ul>
          </div>
        </div>
      </nav>
    </div>
  </header>
  <section
    className="form top-bottom-space"
  >
    <div
      className="container"
    >
      <section
        className="section-format col"
      >
        <form>
          <div
            className="field "
          >
            <label>
              Selecciona un tema
            </label>
            <div
              className="select-simple"
            >
              <div
                className="select-simple__container"
              >
                <select
                  name="categoryCombo"
                  onChange={[Function]}
                  value=""
                >
                  <option
                    value=""
                  >
                    Seleccionar
                  </option>
                  <option
                    value="cats"
                  >
                    Gatos
                  </option>
                  <option
                    value="dogs"
                  >
                    Perros
                  </option>
                  <option
                    value="other"
                  >
                    Otro
                  </option>
                </select>
              </div>
            </div>
          </div>
        </form>
      </section>
    </div>
  </section>
</main>
`;
